diff --git a/packages/services/Telephony/src/com/android/phone/CallNotifier.java b/packages/services/Telephony/src/com/android/phone/CallNotifier.java
index 0d7c70d..89d09b2 100644
--- a/packages/services/Telephony/src/com/android/phone/CallNotifier.java
+++ b/packages/services/Telephony/src/com/android/phone/CallNotifier.java
@@ -318,6 +318,15 @@ public class CallNotifier extends Handler
 
             case CallStateMonitor.PHONE_INCOMING_RING:
                 log("PHONE_INCOMING_RING !");
+                //removeMessages(CallStateMonitor.PHONE_INCOMING_RING_DELAY);
+
+                Message delayMsg = Message.obtain(msg);
+                delayMsg.what = CallStateMonitor.PHONE_INCOMING_RING_DELAY;
+                sendMessageDelayed(delayMsg, CallStateMonitor.PHONE_INCOMING_RING_DEALYT_TIME);
+                break;
+
+            case CallStateMonitor.PHONE_INCOMING_RING_DELAY:
+                log("PHONE_INCOMING_RING_DELAY !");
                 // repeat the ring when requested by the RIL, and when the user has NOT
                 // specifically requested silence.
                 if (msg.obj != null && ((AsyncResult) msg.obj).result != null) {
diff --git a/packages/services/Telephony/src/com/android/phone/CallStateMonitor.java b/packages/services/Telephony/src/com/android/phone/CallStateMonitor.java
index b4e9d17..b6b6b95 100644
--- a/packages/services/Telephony/src/com/android/phone/CallStateMonitor.java
+++ b/packages/services/Telephony/src/com/android/phone/CallStateMonitor.java
@@ -61,6 +61,9 @@ class CallStateMonitor extends Handler {
     public static final int PHONE_RESEND_MUTE = 12;
     public static final int PHONE_ON_DIAL_CHARS = 13;
 
+    public static final int PHONE_INCOMING_RING_DELAY = 102; // Added by Joseth
+    public static final int PHONE_INCOMING_RING_DEALYT_TIME = 2000; // ms
+
     // Other events from call manager
     public static final int EVENT_OTA_PROVISION_CHANGE = 20;
 
